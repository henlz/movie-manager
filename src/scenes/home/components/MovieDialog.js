import React, {Component} from 'react';
import PropTypes from 'prop-types';
import Dialog from 'material-ui/Dialog';
import FlatButton from 'material-ui/FlatButton';
import TextField from 'material-ui/TextField';

import './MovieDialog.css';

class MovieDialog extends Component {
  constructor(props) {
    super(props);

    this.movie = props.movie;

    this.modalTitle = !!props.movie.title ? 'Edit movie' : 'Add movie';

    this.state = {
      open: props.isOpen,
      title: this.movie.title,
      description: this.movie.description,
      imageData: this.movie.imageData
    };
  }

  componentWillReceiveProps(props) {
    if (this.props.isOpen !== props.isOpen) {
      this.movie = props.movie;

      this.setState({
        open: props.isOpen,
        title: this.movie.title,
        description: this.movie.description,
        imageData: this.movie.imageData
      });
    }
  }

  handleClose = () => {
    this.props.onClose();
  };

  onChangeTitle = event => {
    this.setState({
      title: event.target.value
    });
  };

  onChangeDescription = event => {
    this.setState({
      description: event.target.value
    });
  };

  handleImageChange = event => {
    event.preventDefault();

    let reader = new FileReader();
    let file = event.target.files[0];

    reader.onloadend = () => {
      this.setState({
        imageData: reader.result
      });
    };

    reader.readAsDataURL(file);
  };

  handleSubmit = () => {
    const movie = {
      id: this.props.movie.id,
      title: this.state.title,
      description: this.state.description,
      imageData: this.state.imageData
    };

    this.props.onSave(movie);
    this.handleClose();
  };

  render() {
    const actions = [
      <FlatButton
        label="Cancel"
        primary={true}
        onTouchTap={this.handleClose}
      />,
      <FlatButton
        label="Submit"
        primary={true}
        keyboardFocused={true}
        onTouchTap={this.handleSubmit}
      />,
    ];

    return (
      <Dialog title={this.modalTitle}
              actions={actions}
              modal={false}
              open={this.state.open}
              onRequestClose={this.handleClose}>

        <div className="movie-text-inputs">
          <img className="movie-image-preview"
               src={this.state.imageData}
               alt={this.props.movie.title}/>
          <input type="file" onChange={this.handleImageChange}/>
        </div>

        <div className="movie-text-inputs">
          <TextField floatingLabelText="Movie title"
                     value={this.state.title}
                     onChange={this.onChangeTitle}/>

          <TextField floatingLabelText="Movie description"
                     value={this.state.description}
                     onChange={this.onChangeDescription}/>
        </div>
      </Dialog>
    );
  }
}

MovieDialog.propTypes = {
  buttonText: PropTypes.string.isRequired,
  movie: PropTypes.object,
  onSave: PropTypes.func.isRequired,
  onClose: PropTypes.func,
  isOpen: PropTypes.bool
};

MovieDialog.defaultProps = {
  movie: {}
};

export default MovieDialog;
