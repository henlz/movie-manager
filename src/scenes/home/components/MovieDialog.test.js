import React from 'react';
import renderer from 'react-test-renderer';
import {mount} from 'enzyme';
import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';

import MovieDialog from './MovieDialog';

describe('MovieDialog scene', function() {
  let component, wrapper, onCloseSpy, onSaveSpy;

  beforeEach(() => {
    onCloseSpy = jest.fn();
    onSaveSpy = jest.fn();

    component = (
      <MuiThemeProvider>
        <MovieDialog buttonText='Text'
                     onSave={onSaveSpy}
                     onClose={onCloseSpy}
                     isOpen={false}/>
      </MuiThemeProvider>
    );

    wrapper = mount(component);
  });

  it('renders and matchs snapshot', function() {
    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should set an Add title', function() {
    const dialogNode = wrapper.find('Dialog').node;

    expect(dialogNode.props.title).toBe('Add movie');
  });

  it('should set an Add title', function() {
    const movie = {
      title: 'movie title'
    };
    component = (
      <MuiThemeProvider>
        <MovieDialog buttonText='Text'
                     movie={movie}
                     onSave={() => false}
                     isOpen={true}/>
      </MuiThemeProvider>
    );

    wrapper = mount(component);

    const dialogNode = wrapper.find('Dialog').node;

    expect(dialogNode.props.title).toBe('Edit movie');
  });

  it('handleClose should call onClose prop', function() {
    const movieDialogNode = wrapper.find('MovieDialog').node;
    movieDialogNode.handleClose();

    expect(onCloseSpy.mock.calls.length).toBe(1);
  });

  it('onChangeDescription should set state with description', function() {
    const text = 'text';
    const event = {
      target: {
        value: text
      }
    };
    const movieDialogNode = wrapper.find('MovieDialog').node;

    movieDialogNode.onChangeDescription(event);

    expect(movieDialogNode.state.description).toBe(text);
  });

  it('onChangeTitle should set state with title', function() {
    const text = 'text';
    const event = {
      target: {
        value: text
      }
    };
    const movieDialogNode = wrapper.find('MovieDialog').node;

    movieDialogNode.onChangeTitle(event);

    expect(movieDialogNode.state.title).toBe(text);
  });

  it('handleClose should call onClose prop', function() {
    const movieDialogNode = wrapper.find('MovieDialog').node;
    movieDialogNode.handleClose();

    expect(onCloseSpy.mock.calls.length).toBe(1);
  });
});
