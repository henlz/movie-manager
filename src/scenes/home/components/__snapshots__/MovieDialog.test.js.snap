// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieDialog scene renders and matchs snapshot 1`] = `null`;
